File: probfilter-core/src/main/scala/com/c6h5no2/probfilter/crdt/CausalHistory.scala
package com.c6h5no2.probfilter.crdt

import com.c6h5no2.probfilter.util.UnsignedNumber


trait CausalHistory {
  /**
   * @param replicaId 16-bit unsigned replica id
   * @return 32-bit unsigned timestamp
   */
  def get(replicaId: Short): Int

  /** @note This is a query operation. */
  def next(replicaId: Short): Int = get(replicaId) + 1

  /** @note The default implementation assumes consecutive timestamps; override if necessary. */
  def observes(replicaId: Short, timestamp: Int): Boolean = UnsignedNumber.compare(timestamp, get(replicaId)) <= 0

  def increase(replicaId: Short): CausalHistory
}

File: probfilter-core/src/main/java/com/c6h5no2/probfilter/crdt/CausalHistory.java
package probfilter.crdt;

import probfilter.util.UnsignedNumber;

import java.io.Serializable;


public interface CausalHistory extends Serializable {
    /**
     * @param replicaId 16-bit unsigned replica id
     * @return 32-bit unsigned timestamp
     */
    int get(short replicaId);

    default int next(short replicaId) {
        return get(replicaId) + 1;
    }

    /**
     * @implNote The default implementation assumes consecutive timestamps; override if necessary.
     */
    default boolean observes(short replicaId, int timestamp) {
        return UnsignedNumber.compare(timestamp, get(replicaId)) <= 0;
    }

    CausalHistory increase(short replicaId);
}
